vault/policy.go

```diff
@@ -8,64 +8,36 @@ import (
 )
 
 const (
-	PathPolicyDeny  = "deny"
-	PathPolicyRead  = "read"
-	PathPolicyWrite = "write"
-	PathPolicySudo  = "sudo"
+	CreateCapability = "create"
+	ReadCapability   = "read"
+	UpdateCapability = "update"
+	DeleteCapability = "delete"
+	ListCapability   = "list"
+	DenyCapability   = "deny"
+	SudoCapability   = "sudo"
+
+	// Backwards compatibility
+	OldDenyPathPolicy  = "deny"
+	OldReadPathPolicy  = "read"
+	OldWritePathPolicy = "write"
+	OldSudoPathPolicy  = "sudo"
 )
 
 // Policy is used to represent the policy specified by
 // an ACL configuration.
 type Policy struct {
-	Name  string        `hcl:"name"`
-	Paths []*PathPolicy `hcl:"path,expand"`
+	Name  string              `hcl:"name"`
+	Paths []*PathCapabilities `hcl:"path,expand"`
 	Raw   string
 }
 
-// PathPolicy represents a policy for a path in the namespace
-type PathPolicy struct {
-	Prefix string `hcl:",key"`
-	Policy string
-	Glob   bool
-}
-
-// TakesPrecedence is used when multiple policies
-// collide on a path to determine which policy takes
-// precendence.
-func (p *PathPolicy) TakesPrecedence(other *PathPolicy) bool {
-	// Handle the full merge matrix
-	switch p.Policy {
-	case PathPolicyDeny:
-		// Deny always takes precendence
-		return true
-
-	case PathPolicyRead:
-		// Read never takes precedence
-		return false
-
-	case PathPolicyWrite:
-		switch other.Policy {
-		case PathPolicyRead:
-			return true
-		case PathPolicyDeny, PathPolicyWrite, PathPolicySudo:
-			return false
-		default:
-			panic("missing case")
-		}
-
-	case PathPolicySudo:
-		switch other.Policy {
-		case PathPolicyRead, PathPolicyWrite:
-			return true
-		case PathPolicyDeny, PathPolicySudo:
-			return false
-		default:
-			panic("missing case")
-		}
-
-	default:
-		panic("missing case")
-	}
+// Capability represents a policy for a path in the namespace
+type PathCapabilities struct {
+	Prefix          string `hcl:",key"`
+	Policy          string
+	Capabilities    []string
+	CapabilitiesMap map[string]bool `hcl:"-"`
```

Comment:
Since this is internal, lets make it a uint32 bitmap and avoid the map allocation and hash lookups.